File: pixel/frontend/src/constants/thresholds.ts
// Sensor warning/critical thresholds for 2026 Honda Accord SE L15BE 1.5T
// Sources: Honda service manual operating ranges, OBD-II standard ranges
// Values verified against Honda Accord community data and service specs

export interface SensorThreshold {
  warn: { low?: number; high?: number };
  critical: { low?: number; high?: number };
  unit: string;
  label: string;
}

export const SENSOR_THRESHOLDS: Record<string, SensorThreshold> = {
  RPM: {
    label: "RPM",
    unit: "rpm",
    warn: { high: 5500 },
    critical: { high: 6500 },
  },
  SPEED: {
    label: "Speed",
    unit: "kph",
    warn: {},
    critical: {},
  },
  COOLANT_TEMP: {
    // Backend: normal 70-100, warning 40-105, critical 0-110
    label: "Coolant",
    unit: "\u00B0C",
    warn: { high: 105, low: 40 },
    critical: { high: 110, low: 0 },
  },
  ENGINE_LOAD: {
    label: "Load",
    unit: "%",
    warn: { high: 85 },
    critical: { high: 95 },
  },
  THROTTLE_POS: {
    label: "Throttle",
    unit: "%",
    warn: {},
    critical: {},
  },
  MAF: {
    label: "MAF",
    unit: "g/s",
    warn: {},
    critical: {},
  },
  FUEL_LEVEL: {
    label: "Tank",
    unit: "%",
    warn: { low: 15 },
    critical: { low: 8 },
  },
  BATTERY_V: {
    // Backend: normal 12.0-15.0, warning 11.8-15.2, critical 11.0-16.0
    // Honda ELD cycles voltage to 12.4-12.9V during cruising -- that is normal
    label: "Volts",
    unit: "V",
    warn: { low: 11.8, high: 15.2 },
    critical: { low: 11.0, high: 16.0 },
  },
  STFT: {
    label: "STFT",
    unit: "%",
    warn: { low: -15, high: 15 },
    critical: { low: -25, high: 25 },
  },
  LTFT: {
    // Backend: normal +/-5%, warning +/-8%, critical +/-10%
    label: "LTFT",
    unit: "%",
    warn: { low: -8, high: 8 },
    critical: { low: -10, high: 10 },
  },
  INTAKE_TEMP: {
    label: "Intake",
    unit: "\u00B0C",
    warn: { high: 55 },
    critical: { high: 65 },
  },
  MAP: {
    label: "MAP",
    unit: "kPa",
    warn: {},
    critical: {},
  },
  CATALYST_TEMP: {
    // Backend: normal 300-800, warning 100-950, critical 0-1050
    label: "Cat",
    unit: "\u00B0C",
    warn: { high: 950 },
    critical: { high: 1050 },
  },
  OIL_TEMP: {
    // Backend: normal 80-120, warning 40-130, critical 0-140
    label: "Oil",
    unit: "\u00B0C",
    warn: { high: 130, low: 40 },
    critical: { high: 140, low: 0 },
  },
  CVT_TEMP: {
    // Backend: normal 50-100, warning 20-115, critical 0-130
    label: "CVT",
    unit: "\u00B0C",
    warn: { high: 115 },
    critical: { high: 130 },
  },
};

export type SensorState = "normal" | "warn" | "critical";

// Returns "normal" | "warn" | "critical" for a sensor value
export function getSensorState(
  key: string,
  value: number,
): SensorState {
  const t = SENSOR_THRESHOLDS[key];
  if (!t) return "normal";

  const { critical, warn } = t;

  if (
    (critical.high !== undefined && value >= critical.high) ||
    (critical.low !== undefined && value <= critical.low)
  ) {
    return "critical";
  }

  if (
    (warn.high !== undefined && value >= warn.high) ||
    (warn.low !== undefined && value <= warn.low)
  ) {
    return "warn";
  }

  return "normal";
}

// Returns "normal" | "warn" | "critical" for overall health score
export function getHealthState(score: number): SensorState {
  if (score === -1) return "normal"; // calibrating
  if (score < 50) return "critical";
  if (score < 70) return "warn";
  return "normal";
}

// Sensor label from thresholds config
export function getSensorLabel(key: string): string {
  return SENSOR_THRESHOLDS[key]?.label ?? key;
}

// Sensor unit from thresholds config
export function getSensorUnit(key: string): string {
  return SENSOR_THRESHOLDS[key]?.unit ?? "";
}

// Find the worst-scoring subsystem from HealthScores
export function getWorstSubsystem(
  health: import("@/types/vehicle").HealthScores,
): { id: import("@/types/vehicle").SubsystemId; score: number } | null {
  const subsystems: import("@/types/vehicle").SubsystemId[] = [
    "engine", "transmission", "fuel", "cooling", "exhaust", "electrical",
  ];

  let worst: { id: import("@/types/vehicle").SubsystemId; score: number } | null = null;

  for (const id of subsystems) {
    const score = health[id];
    if (score === -1) continue; // skip calibrating
    if (worst === null || score < worst.score) {
      worst = { id, score };
    }
  }

  return worst;
}

File: pi/backend/health/thresholds.py
"""Honda Accord 2026 SE (L15BE 1.5T) normal/warning/critical operating ranges.

Values sourced from Honda service manual data, forum scan tool logs, and
ISO 15031-5 PID specifications. Verified against real-world OBD-II captures
from 10th/11th gen Accord 1.5T owners.

Each parameter has three zones:
- Normal: 100 points, everything is fine
- Warning: 85 points, worth watching
- Critical: 50 points, needs attention soon

Important Honda-specific notes:
- Honda ELD (Electrical Load Detector) intentionally drops alternator output
  to 12.4-12.9V during low-load cruising to reduce fuel consumption. This is
  NORMAL behavior, not a charging system fault. The voltage thresholds here
  account for this by keeping the normal floor at 12.0V.
- Oil temp on L15BE is ECU-calculated (no physical oil temp sensor). Values
  come from PID 015C which returns the ECU's estimate.
- Catalyst runs cooler on the 1.5T than NA engines because the turbo has an
  integrated exhaust manifold that absorbs heat before the cat.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class RangeThreshold:
    """Threshold for a parameter with a normal range (low-high)."""
    normal_low: float
    normal_high: float
    warning_low: float
    warning_high: float
    critical_low: float
    critical_high: float


@dataclass(frozen=True)
class AbsoluteThreshold:
    """Threshold for a parameter where deviation from zero matters (fuel trims)."""
    normal: float      # absolute value within this is normal
    warning: float     # absolute value within this is warning
    critical: float    # absolute value beyond this is critical


# --- Honda Accord 2026 SE (L15BE 1.5T) Operating Ranges ---

# Coolant temp: thermostat opens at ~82C, stabilizes 88-95C in normal driving.
# City driving in summer heat can push to 100-103C -- still within Honda spec.
# Warning starts at 100C. Critical at 108C (Honda fans go high-speed at ~107C).
# Low-side thresholds: during warmup, temps below normal_low are expected.
# warning_low acts as a "still warming up" floor.
# critical_low is the genuine concern level (sensor failure or extreme cold).
COOLANT_TEMP = RangeThreshold(
    normal_low=75, normal_high=100,
    warning_low=40, warning_high=105,
    critical_low=0, critical_high=110,
)

# Oil temp: ECU-calculated, not a physical sensor on L15BE.
# Normal operating range 80-120C. Long highway pulls or hot summer city
# driving can sustain 110-115C without issue.
OIL_TEMP = RangeThreshold(
    normal_low=80, normal_high=120,
    warning_low=40, warning_high=130,
    critical_low=0, critical_high=140,
)

# Idle RPM (Park/Neutral, engine warm): Honda targets 700 RPM.
# With AC compressor cycling, can sit 750-850. Below 550 or above 1000
# at warm idle indicates an issue.
IDLE_RPM = RangeThreshold(
    normal_low=700, normal_high=850,
    warning_low=550, warning_high=1000,
    critical_low=400, critical_high=1200,
)

# Battery voltage: Honda uses ELD to intentionally cycle alternator output.
# Normal driving sees 12.4-14.8V depending on ELD state. A sustained reading
# of 12.0V is the low end of acceptable. Below 11.8V or above 15.2V is a
# real charging issue.
BATTERY_VOLTAGE = RangeThreshold(
    normal_low=12.0, normal_high=15.0,
    warning_low=11.8, warning_high=15.2,
    critical_low=11.0, critical_high=16.0,
)

# Engine load at idle: 15-40% is normal. AC compressor adds ~10-15%.
# 25-40% with AC on is expected. Above 50% sustained at idle is unusual.
ENGINE_LOAD_IDLE = RangeThreshold(
    normal_low=15, normal_high=40,
    warning_low=10, warning_high=50,
    critical_low=5, critical_high=60,
)

# CVT fluid temp: Honda CVT fluid operates cooler than traditional ATF.
# Normal range 50-100C. Mountain towing or aggressive city driving can
# push 100-115C. Above 130C risks CVT belt/chain damage.
CVT_FLUID_TEMP = RangeThreshold(
    normal_low=50, normal_high=100,
    warning_low=20, warning_high=115,
    critical_low=0, critical_high=130,
)

# Catalyst temp: turbo integrated exhaust manifold absorbs heat before the
# cat, so the 1.5T runs cooler than NA engines. Normal operating range
# 300-800C. Sustained above 1000C indicates a problem (misfires, rich
# running dumping unburned fuel into the cat).
CATALYST_TEMP = RangeThreshold(
    normal_low=300, normal_high=800,
    warning_low=100, warning_high=950,
    critical_low=0, critical_high=1050,
)

# Short-term fuel trim: +/-10% is normal closed-loop correction.
# +/-10 to +/-15% means the ECU is working harder to compensate.
# Beyond +/-25% the ECU is near its correction limit.
STFT = AbsoluteThreshold(normal=10.0, warning=15.0, critical=25.0)

# Long-term fuel trim: +/-5% is well-adapted. +/-5 to +/-8% means slow
# drift (dirty air filter, slight vacuum leak). Beyond +/-10% sustained
# means the ECU is near its adaptation limit.
# Gap between warning and critical prevents score oscillation from noise.
LTFT = AbsoluteThreshold(normal=5.0, warning=8.0, critical=10.0)


# --- EWMA alpha values ---
# Empirically grounded based on sensor thermal mass and update characteristics.
# alpha = 2/(N+1) where N is the effective window in samples at 10Hz.
EWMA_ALPHAS: dict[str, float] = {
    "coolant_temp_c": 0.01,     # slow thermal mass, N~200, ~20s window
    "ltft_pct": 0.02,           # very slow ECU-averaged adaptation, N~100, ~10s
    "battery_voltage": 0.05,    # fast but ELD cycles, N~40, ~4s to smooth cycling
    "catalyst_temp_c": 0.02,    # moderate thermal mass, N~100, ~10s
    "oil_temp_c": 0.01,         # ECU-calculated, slow, N~200, ~20s
}


# --- Subsystem to PID mapping ---
# Which VehicleSnapshot fields feed into each subsystem's health score

SUBSYSTEM_PIDS: dict[str, list[str]] = {
    "engine":       ["rpm", "engine_load_pct", "throttle_pct", "maf_gps", "intake_manifold_kpa"],
    "transmission": ["cvt_fluid_temp_c"],
    "fuel":         ["stft_pct", "ltft_pct"],
    "cooling":      ["coolant_temp_c", "oil_temp_c"],
    "exhaust":      ["catalyst_temp_c"],
    "electrical":   ["battery_voltage"],
}

# How much each subsystem contributes to the overall score
SUBSYSTEM_WEIGHTS: dict[str, float] = {
    "engine":       0.30,
    "transmission": 0.20,
    "fuel":         0.15,
    "cooling":      0.15,
    "exhaust":      0.10,
    "electrical":   0.10,
}

# DTC severity penalties (applied to the relevant subsystem)
DTC_PENALTY_GENERIC = 15    # confirmed generic powertrain code
DTC_PENALTY_CRITICAL = 30   # misfire (P0300-P0304), catalyst (P0420), overtemp (P0217)
